common/crypto/curl-p/ptrit.c

```diff
@@ -1,75 +1,433 @@
+      166, 530, 165, 529, 164, 528, 163, 527, 162, 526, 161, 525, 160, 524, 159, 523, 158, 522, 157, 521, 156, 520,    \
+      155, 519, 154, 518, 153, 517, 152, 516, 151, 515, 150, 514, 149, 513, 148, 512, 147, 511, 146, 510, 145, 509,    \
+      144, 508, 143, 507, 142, 506, 141, 505, 140, 504, 139, 503, 138, 502, 137, 501, 136, 500, 135, 499, 134, 498,    \
+      133, 497, 132, 496, 131, 495, 130, 494, 129, 493, 128, 492, 127, 491, 126, 490, 125, 489, 124, 488, 123, 487,    \
+      122, 486, 121, 485, 120, 484, 119, 483, 118, 482, 117, 481, 116, 480, 115, 479, 114, 478, 113, 477, 112, 476,    \
+      111, 475, 110, 474, 109, 473, 108, 472, 107, 471, 106, 470, 105, 469, 104, 468, 103, 467, 102, 466, 101, 465,    \
+      100, 464, 99, 463, 98, 462, 97, 461, 96, 460, 95, 459, 94, 458, 93, 457, 92, 456, 91, 455, 90, 454, 89, 453, 88, \
+      452, 87, 451, 86, 450, 85, 449, 84, 448, 83, 447, 82, 446, 81, 445, 80, 444, 79, 443, 78, 442, 77, 441, 76, 440, \
+      75, 439, 74, 438, 73, 437, 72, 436, 71, 435, 70, 434, 69, 433, 68, 432, 67, 431, 66, 430, 65, 429, 64, 428, 63,  \
+      427, 62, 426, 61, 425, 60, 424, 59, 423, 58, 422, 57, 421, 56, 420, 55, 419, 54, 418, 53, 417, 52, 416, 51, 415, \
+      50, 414, 49, 413, 48, 412, 47, 411, 46, 410, 45, 409, 44, 408, 43, 407, 42, 406, 41, 405, 40, 404, 39, 403, 38,  \
+      402, 37, 401, 36, 400, 35, 399, 34, 398, 33, 397, 32, 396, 31, 395, 30, 394, 29, 393, 28, 392, 27, 391, 26, 390, \
+      25, 389, 24, 388, 23, 387, 22, 386, 21, 385, 20, 384, 19, 383, 18, 382, 17, 381, 16, 380, 15, 379, 14, 378, 13,  \
+      377, 12, 376, 11, 375, 10, 374, 9, 373, 8, 372, 7, 371, 6, 370, 5, 369, 4, 368, 3, 367, 2, 366, 1, 365, 0
+
+static size_t const curl_index[CURL_STATE_SIZE + 1] = {CURL_SBOX_INDEX_TABLE};
+// 0, 364, 728, 363, 727, ..., 2, 366, 1, 365, 0
+#endif  // PCURL_S2_ARGS_LUT
+
+#if defined(PCURL_STATE_DOUBLE)
+static FORCED_INLINE void pcurl_sbox(ptrit_t *c, ptrit_t const *s) {
+#if defined(PCURL_S2_ARGS_LUT)
+  size_t i;
+
+#if defined(PCURL_SBOX_UNWIND_2)
+  pcurl_s2(s + curl_index[0], s + curl_index[1], c);
+  for (i = 1; i < CURL_STATE_SIZE;) {
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
+  }
+#elif defined(PCURL_SBOX_UNWIND_4)
+  pcurl_s2(s + curl_index[0], s + curl_index[1], c);
+  for (i = 1; i < CURL_STATE_SIZE;) {
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
+  }
+#elif defined(PCURL_SBOX_UNWIND_8)
+  pcurl_s2(s + curl_index[0], s + curl_index[1], c);
+  for (i = 1; i < CURL_STATE_SIZE;) {
+    pcurl_s2(s + curl_index[i], s + curl_index[i + 1], c + i);
+    ++i;
```

Comment:
can you export this code into another loop and set number of iterations based on #define PCURL_SBOX_* value?

Reply:
The whole idea was to unwind these loops. Of course we can put them into an internal loop, but will the compiler figure out that this loop needs to be unwound?

Reply:
My bad then, 
but seems like there is a gcc option:
https://stackoverflow.com/questions/4071690/tell-gcc-to-specifically-unroll-a-loop/14251545

Reply:
I'm not sure how to use `__attribute__((optimize("unroll-loops")))`. Does it unroll the loop completely? But I want to unroll just a few rounds. Plus it's gcc-specific. IMO it's easier to just drop support for PCURL_SBOX_UNWIND, the speedup gain seems to be very small - a few percents.